### ISSUE TITLE
[General Issue] 关于SemanticConnectivityLoss是否可导

### ISSUE THREAD

**Issue by eezywu** [PaddlePaddle/PaddleSeg#2695]
您好！关于SemanticConnectivityLoss我有个疑问，就是[代码第65行](https://github.com/PaddlePaddle/PaddleSeg/blob/release/2.6/paddleseg/models/losses/semantic_connectivity_loss.py#L65)
 `preds = paddle.argmax(logits, axis=1) if self.use_argmax else logits`使用了argmax，此处由于preds是从argmax得到的，而argmax是不可导操作，那理论上preds之后的运算应该不会产生梯度回传到logits了吧？希望您能解答一下，感谢！

**Comment by juncaipeng** [comment]
由于SemanticConnectivityLoss应用于人像这种两类分割问题，该loss的输入应该是只有一个通道，loss实现中也支持设置不使用argmax。

**Comment by eezywu** [comment]
> 由于SemanticConnectivityLoss应用于人像这种两类分割问题，该loss的输入应该是只有一个通道，loss实现中也支持设置不使用argmax。

我看后面对pred求连通域/iou等操作，应该是要求pred中的值为class_id，所以即使输入只有一个通道，我理解也需要将其二值化，而二值化操作也是不可导操作

**Comment by eezywu** [comment]
您好！请问这个问题有结论吗？现在还是有些困扰，希望能帮忙解答一下哈，非常感谢！

**Comment by LutaoChu** [comment]
你好，argmax是可以回传梯度的，类似于max_pooling，会将梯度传回到最大元素的位置，其他元素梯度为0.  早起分类网络大量使用max_pooling，如果不可导都没法训练了

**Comment by LutaoChu** [comment]
另外，求取连通性用的numpy API，是没有反向算子的。为了解决这个问题，我用了一个trick，在这里mask和pred相乘提取连通区域，将梯度传回 pred   https://github.com/PaddlePaddle/PaddleSeg/blob/875eca5911ac4679b5b399a1af9890d74eec4d1f/paddleseg/models/losses/semantic_connectivity_loss.py#L146

**Comment by eezywu** [comment]
> 

感谢回复！我理解应该和max_pooling不太一样，max_pooling可以copy输入tensor的属性，但从paddle源码[这里](https://github.com/PaddlePaddle/Paddle/blob/2739bd733e81f288b62797f28e53bb8465e2dbc3/paddle/phi/kernels/cpu/viterbi_decode_kernel.cc#L69)可以看到argmax取的是输入tensor的index（index没有图连接），并且[这里](https://github.com/PaddlePaddle/Paddle/tree/2739bd733e81f288b62797f28e53bb8465e2dbc3/paddle/phi/kernels/cpu)似乎也没有实现argmax的梯度算子。期待您的回复！

**Comment by sunjian2015** [comment]
> 你好，argmax是可以回传梯度的，类似于max_pooling，会将梯度传回到最大元素的位置，其他元素梯度为0. 早起分类网络大量使用max_pooling，如果不可导都没法训练了

请问这个loss有pytorch的实现吗？我自己改了一版本，但是由于数据在gpu与cpu、是否需要梯度，出现各种问题....
